### PR TITLE
Update Redux-Saga and add flow-typed as well as own typing layer

### LINKED ISSUE #2775: Upgrade redux-saga and integrate flow-types for redux-saga

### LINKED ISSUE #2966: Improve saga typings

### PR DESCRIPTION
**Pull request by philippotto**
I upgraded redux-saga to 0.16 and added the official flow-typed definitions. For that I needed to adapt some things, as we didn't adhere strictly to the official interface. Additionally, I added a thin, typed wrapper for redux functions, which change the interface slightly, but allow for greatly improved typing.

The difference in usage is mainly that we have to use `yield* selectNew(...)` instead of `yield selectOld(...)`. `selectNew` is the `select` function from the new wrapper and `selectOld` is the conventional one provided by redux-saga. When we use something like `call(...)` in a saga without yielding it directly (e.g., because it's passed to a `race` effect), we have to use the conventional `call` method, unfortunately. This is a bit cumbersome and we should discuss how to approach this while reducing confusion.
Within tests, the standard effects are still used, because we only care about the descriptors there.

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- [ ] test protocol?

### Issues:
- fixes #2775 
- fixes #2966

------
- [ ] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable
- [ ] Ready for review

### CODE REVIEW

File: app/assets/javascripts/admin/admin_rest_api.js [hunk 1]
```diff
@@ -427,7 +427,7 @@ export function editAnnotation(
   annotationId: string,
   annotationType: APITracingType,
   data: $Shape<EditableAnnotationType>,
-): Promise<void> {
+): Promise<any> {
```
Comment (daniel-wer) [inline app/assets/javascripts/admin/admin_rest_api.js]: The any can probably be replaced by a more specific type or does this pose problems?

File: app/assets/javascripts/oxalis/model/sagas/save_saga.js [hunk 2]
```diff
@@ -166,29 +177,29 @@ export function* sendRequestToServer(
...
     }
-    yield delay(SAVE_RETRY_WAITING_TIME);
-    yield call(sendRequestToServer);
+    yield* call(delay, SAVE_RETRY_WAITING_TIME);
+    yield* call(sendRequestToServer);
   }
 }
 
-export function toggleErrorHighlighting(state: boolean) {
+export function toggleErrorHighlighting(state: boolean): void {
```
Comment (daniel-wer) [inline app/assets/javascripts/oxalis/model/sagas/save_saga.js]: Isn't `void` the implicit return type?

File: app/assets/javascripts/oxalis/model/sagas/annotation_saga.js [hunk 3]
```diff
@@ -12,15 +15,16 @@ import {
...
-  const tracing = yield select(state => state.tracing);
-  yield [
-    call(editAnnotation, tracing.annotationId, tracing.tracingType, {
-      name: tracing.name,
-      isPublic: tracing.isPublic,
-      description: tracing.description,
-    }),
-  ];
+export function* pushAnnotationUpdateAsync(): Saga<void> {
+  const tracing = yield* _select(state => state.tracing);
```
Comment (daniel-wer) [inline app/assets/javascripts/oxalis/model/sagas/annotation_saga.js]: Just a reminder to rename the `_select` and `_call` according to how you did it in the save_saga :)

**Comment by philippotto** [comment]
For the record, the overall flow coverage only increased by one percent point, but within the (quite mission-critical) sagas, the flow coverage increased from below 30% to above 95%.

Before:
image

After:
image


I don't understand why the coverage in the root saga is relatively low (the regions marked as "not typed" look typed to me), but in the end, the exact numbers don't matter much, anyway.

File: app/assets/javascripts/oxalis/model/sagas/save_saga.js (hunk 2 above)
Comment (philippotto) [reply app/assets/javascripts/oxalis/model/sagas/save_saga.js]: We already talked about this, but for the record: Flow wasn't able to infer this for some reason and complained here.

**Pull request merged**